core/domain/exp_domain.py

```diff
@@ -3519,3 +3519,226 @@ def to_metadata_dict(self):
+        model = cls.get_state_id_mapping_from_model(
+            exp_models.StateIdMappingModel.get_state_id_mapping_model(
+                old_exploration.id, old_exploration.version, strict=True))
+
+        # Generate state id to name mapping dict for old exploration.
+        for state_name in old_exploration.states:
+            state_id_to_names[model.get_state_id(state_name)] = state_name
+
+        old_state_id_to_names = copy.deepcopy(state_id_to_names)
+
+        # Analyse each command in change list one by one to create state id
+        # mapping for new exploration.
+        for change_dict in change_list:
+            if change_dict['cmd'] == CMD_ADD_STATE:
+                new_state_names.append(change_dict['state_name'])
+
+            elif change_dict['cmd'] == CMD_DELETE_STATE:
+                removed_state_name = change_dict['state_name']
+
+                # Find state name in state id mapping. If it exists then
+                # remove it from mapping.
+                if removed_state_name in state_id_to_names.values():
+                    key = None
+                    for (state_id, state_name) in state_id_to_names.iteritems():
+                        if state_name == removed_state_name:
+                            key = state_id
+                            break
+                    state_id_to_names.pop(key)
+                else:
+                    # If the state is not present in state id mapping then
+                    # remove it from new state names list.
+                    new_state_names.remove(removed_state_name)
+
+            elif change_dict['cmd'] == CMD_RENAME_STATE:
+                old_state_name = change_dict['old_state_name']
+                new_state_name = change_dict['new_state_name']
+
+                # Find whether old state name is present in state id mapping.
+                # If it is then replace the old state name to new state name.
+                if old_state_name in state_id_to_names.values():
+                    key = None
+                    for (state_id, state_name) in state_id_to_names.iteritems():
+                        if state_name == old_state_name:
+                            key = state_id
+                            break
+                    state_id_to_names[key] = new_state_name
+                else:
+                    # If old state name is not present in state id mapping then
+                    # remove it from new state names and add new state name
+                    # in new state names list.
+                    new_state_names.remove(old_state_name)
+                    new_state_names.append(new_state_name)
+
+            elif change_dict['cmd'].endswith('_revert_version_number'):
+                # If command is to revert exploration then obtain state id
+                # mapping for reverted version of exploration.
+                reverted_version = change_dict['version_number']
```

Comment:
For this case, you could check it "outside" the domain object (see my comment above re refactoring). Only after you've ensured it's not a "revert" commit would you pass it to the domain object to do the computations.

Reply:
Done. Took it out in exp_services.